triggers/daml/ClearingTrigger.daml

```diff
@@ -146,3 +162,52 @@ collectDeposits restFunds (used, (nextDeposit::remaining))
                                 ((nextDeposit :: used), remaining)
   where
     nextAmount  = (.asset.quantity) . snd $ nextDeposit
+
+
+-- |Calculates the FairValue for all Derivatives known to Exchange
+handleCalculationRequests : Party -> ContractPair Clearing.FairValueCalculationRequest -> TriggerA ClearingState ()
+handleCalculationRequests party (requestCid, rq@Clearing.FairValueCalculationRequest{..}) = do
+  debug "handling calculation requests..."
+  allListings <- query @Listing.T
+  let optListingIdSet = Set.fromList <$> optListingIds
+  let listings = case optListingIdSet of
+        (Some ids) -> filterField (.listingId) (`Set.member` ids) allListings
+        None       -> allListings
+
+  forA_ listings (handleFairValueCalculation party currency upTo)
+  emitExerciseCmd requestCid Clearing.FairValueCalculationRequest_Ack
+  return ()
+
+-- |Calculates a FairValue for a single derivative.
+-- Order of precedence:
+--    1. The price of the most recent trade before the specified 'upTo'
+--    2. If there are no trades, the most recent 'FairValue'
+--    3. If there are no 'FairValue', request manual calculation
+--
+-- TODO: Allow user to specify a 'from' Time interval before which trades and
+-- fairvalues will be ignored.
+handleFairValueCalculation : Party -> Id -> Time -> ContractPair Listing.T -> TriggerA ClearingState ()
+handleFairValueCalculation party currency upTo (marketPairCid,l@(Listing.Listing{..})) = do
+  debug ("calculating FVR: " <> l.listingId)
+
+  fairValues <- sortOn ((.upTo) . snd)
+            <$> filterQuery @Clearing.FairValue (.listingId) (== l.listingId)
```

Comment:
Btw not sure this `filterQuery` is helping at all; think that in-lining this with a lambda would be much more legible.

Reply:
I personally like the way it looks since if you read through the lambdas it reads as an expression (`Template.fieldName == variable)` but if you're strongly against it I can change. I find the lambdas pattern matching the tuples all over these triggers ugly, but it's readable in that it is exactly how it looks, so that's fine.